File: 2]_Algorithms/8]_Divide and Counqer/(divide and conquer)in java/quicksort.java
public class quicksort {
    //print array function
    public static void printarr(int arr[]){
        for(int i=0;i<arr.length;i++){
            System.out.print(arr[i]);
        }
    }
    // Quick sort
    public static void quickSort(int arr[],int si,int ei){
        //base case
        if(si>=ei){
            return;
        }
        // implementation
        int pIdx= partition(arr,si,ei);
        quickSort(arr,si,pIdx-1);//left part reccursion
        quickSort(arr, pIdx+1, ei);// right part reccursion
    }
    public static int partition(int arr[],int si,int ei){
        int pivot=arr[ei];
        int i=si-1;
        for(int j=si;j<ei;j++){
            if(arr[j]<=pivot){
            i++;
            //increment and swap
            int temp=arr[j];
            arr[j]=arr[i];
            arr[i]=temp;
            }
        }
        i++;
        int temp=pivot;
        arr[ei]=arr[i];// pivot=arr[i] here pivot = variable  variable values does not reflect in functions (call by value)
        arr[i]=temp;
        return i;
    }

    public static void main(String args[]){
        int arr[]={2,1,3,4};
        quickSort(arr, 0, arr.length-1);
        printarr(arr);
    }


    
}

File: 2]_Algorithms/2]_Sorting/Sorting in C++/quickSort.c
#include <stdio.h>

void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void qsort2(int *a, int n)
{
    if (n == 0)
    {
        printf("Array is empty.");
        return;
    }
    if (n < 2)
    {
        return;
    }
    int pivot, left, right, leftLen, rightLen, *leftArr, *rightArr;
    pivot = *(a);
    left = 1;
    for (int right = 1; right <= n; right++)
    {
        if (*(a + right) < pivot)
        {
            swap((a + left), (a + right));
            left++;
        }
    }
    swap(a, (a + left - 1));
    leftLen = left - 1;
    leftArr = a;
    rightLen = n - left;
    rightArr = (a + left);
    qsort2(rightArr, rightLen);
    qsort2(leftArr, leftLen);
};

int main(void)
{
    int arr1[5] = {2, 4, 1, 5, 3};
    printf("Before Quick Sort:\n");
    for (int i = 0; i < 5; i++)
    {
        printf("%d", arr1[i]);
    }
    qsort2(arr1, 5);
    printf("\nAfter Quick Sort:\n");
    for (int i = 0; i < 5; i++)
    {
        printf("%d", arr1[i]);
    }
    return 0;
}